Firebase/Auth/Source/FIRAuthURLPresenter.m

```diff
@@ -0,0 +1,136 @@
+    @brief Returns a default FIRAuthUIDelegate object.
+    @return The default FIRAuthUIDelegate object.
+ */
++ (id<FIRAuthUIDelegate>)defaultUIDelegate;
+@end
+
+@interface FIRAuthURLPresenter () <SFSafariViewControllerDelegate>
+@end
+
+@implementation FIRAuthURLPresenter {
+  /** @var _callbackMatcher
+      @brief The callback URL matcher for the current presentation, if one is active.
+   */
+  FIRAuthURLCallbackMatcher _Nullable _callbackMatcher;
+
+  /** @var _safariViewController
+      @brief The SFSafariViewController used for the current presentation, if any.
+   */
+  SFSafariViewController *_Nullable _safariViewController;
+
+  /** @var _completion
+      @brief The completion handler for the current presentaion, if one is active.
+      @remarks This variable is also used as a flag to indicate a presentation is active.
+   */
+  FIRAuthURLPresentationCompletion _Nullable _completion;
+}
+
+- (void)presentURL:(NSURL *)URL
+        UIDelegate:(nullable id<FIRAuthUIDelegate>)UIDelegate
+   callbackMatcher:(FIRAuthURLCallbackMatcher)callbackMatcher
+        completion:(FIRAuthURLPresentationCompletion)completion {
+  _callbackMatcher = callbackMatcher;
+  _completion = completion;
+  // If a UIDelegate is not provided.
+  if (!UIDelegate) {
+    id<FIRAuthUIDelegate> delegate = [FIRAuthDefaultUIDelegate defaultUIDelegate];
+    [self presentWebContextWithController:delegate URL:URL];
+    return;
+  }
+  // If a valid UIDelegate is provided.
+  [self presentWebContextWithController:UIDelegate URL:URL];
+}
+
+- (BOOL)canHandleURL:(NSURL *)URL {
+  if (_callbackMatcher(URL)) {
+    _callbackMatcher = nil;
+    [self finishPresentationWithURL:URL error:nil];
+    return YES;
+  }
+  return NO;
+}
+
+/** @fn presentWebContextWithController:URL:
+    @brief Presents a SFSafariViewController or WKWebView to display the contents of the URL
+        provided.
+    @param controller The controller used to present the SFSafariViewController or WKWebView.
+    @param URL The URL to display in the SFSafariViewController or WKWebView.
+ */
+- (void)presentWebContextWithController:(id)controller URL:(NSURL *)URL {
+  if ([SFSafariViewController class]) {
+    if (_safariViewController) {
+      // Unable to start a new presentation on top of modal SFSVC presentation.
+      _completion(nil, [FIRAuthErrorUtils webContextAlreadyPresentedErrorWithMessage:nil]);
+      return;
+    }
+    SFSafariViewController *safariViewController = [[SFSafariViewController alloc] initWithURL:URL];
+    _safariViewController = safariViewController;
+    _safariViewController.delegate = self;
+    [controller presentViewController:safariViewController animated:YES completion:nil];
+    return;
+  }
+}
+
+#pragma mark - SFSafariViewControllerDelegate
+
+- (void)safariViewControllerDidFinish:(SFSafariViewController *)controller {
+  if (controller == _safariViewController) {
+    _safariViewController = nil;
+    [self finishPresentationWithURL:nil
```

Comment:
Please add a TODO to investigate if the callback is called too earlier, i.e., before SFSVC is completely closed, and handle it accordingly.